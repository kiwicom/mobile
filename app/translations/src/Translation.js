// @flow

import * as React from 'react';
import { NativeModules } from 'react-native';
import { Text } from '@kiwicom/react-native-app-shared';

import DefaultVocabulary, { type TranslationKeys } from './DefaultVocabulary';
import CaseTransform, {
  type SupportedTransformationsType,
} from './transformations/CaseTransform';

type Props =
  | {|
      id: TranslationKeys,
      values?: Object,
      textTransform?: SupportedTransformationsType,
    |}
  | {|
      passThrough: ?string | ?number, // may be undefined because of IDX output
      textTransform?: SupportedTransformationsType,
    |};

/**
 * Regular translation using registered keys:
 *
 * ```js
 * <Translation id="SingleHotel.Rating.Stars" />
 * ```
 *
 * So called "pass through" translation (this props is not being translated
 * but passed through the component).
 *
 * ```js
 * <Translation passThrough="★★★★★" />
 * ```
 *
 * It's possible to apply text transformations in both cases:
 *
 * ```js
 * <Translation passThrough="★★★★★" textTransform="uppercase" />
 * <Translation id="SingleHotel.Rating.Stars" textTransform="uppercase" />
 * ```
 */
export default class Translation extends React.Component<Props> {
  // use this property to highlight translations for screenshoting or debugging
  highlightTranslations = false;

  replaceValues = (text: string, values?: Object): string => {
    const VARIABLE_REGEX = /__([0-9]+_)?([a-zA-Z-_]+)__/g;
    return text.replace(
      VARIABLE_REGEX,
      (match, variableIndex, variableName) => {
        // 'variableIndex' is currently being ignored (we don't need it)
        const value = values && values[variableName];
        return value !== undefined ? value : '?';
      },
    );
  };

  render = () => {
    const containerStyle = this.highlightTranslations
      ? {
          backgroundColor: this.props.id ? 'orange' : 'pink',
        }
      : undefined;

    if (this.props.id === undefined) {
      // ID prop is missing so let's assume it's "pass through" translation
      // (property passThrough may be undefined)
      return (
        <Text style={containerStyle}>
          {/* $FlowExpectedError: we do not allow to use 'string' in the 'Text' components but translations are exceptions. */}
          {CaseTransform(
            String(this.props.passThrough),
            this.props.textTransform,
          )}
        </Text>
      );
    }

    const key = this.props.id;
    let translatedString = NativeModules.RNTranslationManager.translate(key);

    if (translatedString === key) {
      // fallback to our dummy vocabulary because native code didn't provide translation
      translatedString = DefaultVocabulary[key];
    }

    if (translatedString === undefined) {
      // Everything failed - native translation even our fallback. This usually
      // means we are using key that does not exist. In this case the only thing
      // we can actually do is to use original key as a text.
      translatedString = key;
    }

    const values = this.props.values || {};
    return (
      <Text style={containerStyle}>
        {/* $FlowExpectedError: we do not allow to use 'string' in the 'Text' components but translations are exceptions. */}
        {CaseTransform(
          this.replaceValues(translatedString, values),
          this.props.textTransform,
        )}
      </Text>
    );
  };
}
