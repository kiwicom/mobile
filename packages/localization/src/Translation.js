// @flow

import * as React from 'react';
import { Text } from '@kiwicom/mobile-shared';
import { translate } from '@kiwicom/rnmodules';

import vocabulary, { type TranslationKeys } from './DefaultVocabulary';
import CaseTransform, {
  type SupportedTransformationsType,
} from './transformations/CaseTransform';
import { getTranslation, replaceValues } from './TranslationHelpers';

// Preload translations
Object.keys(vocabulary).forEach(async key => {
  const nativeKey = 'mobile.' + key;

  const translatedString = await translate(nativeKey);
  if (translatedString != null) {
    vocabulary[key] = getTranslation(translatedString, key);
  }
});

type CommonProps = {|
  testID?: string,
  textTransform?: SupportedTransformationsType,
|};

type Props =
  | {|
      id: TranslationKeys,
      values?: Object,
      ...CommonProps,
    |}
  | {|
      passThrough: ?string | ?number, // may be undefined because of IDX output
      ...CommonProps,
    |};

/**
 * Regular translation using registered keys:
 *
 * ```js
 * <Translation id="SingleHotel.Rating.Stars" />
 * ```
 *
 * So called "pass through" translation (this props is not being translated
 * but passed through the component).
 *
 * ```js
 * <Translation passThrough="★★★★★" />
 * ```
 *
 * It's possible to apply text transformations in both cases:
 *
 * ```js
 * <Translation passThrough="★★★★★" textTransform="uppercase" />
 * <Translation id="SingleHotel.Rating.Stars" textTransform="uppercase" />
 * ```
 */
export default class Translation extends React.Component<Props> {
  // use this property to highlight translations for screenshoting or debugging
  highlightTranslations = false;

  render = () => {
    const containerStyle = this.highlightTranslations
      ? {
          backgroundColor: this.props.id ? 'orange' : 'pink',
        }
      : undefined;

    if (this.props.id === undefined) {
      // ID prop is missing so let's assume it's "pass through" translation
      // (property passThrough may be undefined)
      return (
        <Text style={containerStyle}>
          {/* $FlowExpectedError: we do not allow to use 'string' in the 'Text' components but translations are exceptions. */}
          {CaseTransform(
            String(this.props.passThrough),
            this.props.textTransform,
          )}
        </Text>
      );
    }

    const values = this.props.values || {};
    return (
      <Text style={containerStyle}>
        {/* $FlowExpectedError: we do not allow to use 'string' in the 'Text' components but translations are exceptions. */}
        {CaseTransform(
          replaceValues(vocabulary[this.props.id], values),
          this.props.textTransform,
        )}
      </Text>
    );
  };
}
