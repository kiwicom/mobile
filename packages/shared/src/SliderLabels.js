// @flow

import * as React from 'react';
import { defaultTokens } from '@kiwicom/mobile-orbit';
import { View } from 'react-native';

import type { OnLayout } from '../types/Events';
import Text from './Text';
import StyleSheet from './PlatformStyleSheet';
import Price from './Price';
import type { TranslationType } from '../types/Translation';

type Props = {|
  +startLabel: TranslationType | React.Element<typeof Price>,
  +startValue: number,
  +endLabel?: TranslationType | React.Element<typeof Price>,
  +endValue?: number,
  +max: number,
  +min: number,
|};

type State = {|
  width: number,
  labelStartWidth: number,
  labelStartAtMax: boolean,
  labelEndWidth: number,
  paddingLeft: number,
  paddingRight: number,
|};

export default class SliderLabels extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = {
      width: 0,
      labelStartWidth: 0,
      labelStartAtMax: false,
      labelEndWidth: 0,
      paddingLeft: 0,
      paddingRight: 0,
    };
  }

  componentDidMount() {
    if (this.props.endValue) {
      requestAnimationFrame(this.setPaddingForTwoLabels);
    } else {
      requestAnimationFrame(this.setPaddingForOneLabel);
    }
  }

  getMaxPadding = (gap: number): number => {
    return Math.floor(
      this.state.width -
        this.state.labelStartWidth -
        this.state.labelEndWidth -
        gap,
    );
  };

  calculateMarkerStartOffset = (): number => {
    const { min, max, startValue } = this.props;

    let val;
    if (startValue > max) {
      val = max;
    } else if (startValue < min) {
      val = min;
    } else {
      val = startValue;
    }

    return Math.round((val / (max - min)) * this.state.width);
  };

  calculateMarkerEndOffset = (): number => {
    const { min, max, endValue } = this.props;

    if (!endValue) {
      return 0;
    }

    let val;
    if (endValue > max) {
      val = max;
    } else if (endValue < min) {
      val = min;
    } else {
      val = endValue;
    }

    const w = (val / (max - min)) * this.state.width;
    return Math.round(this.state.width - w);
  };

  getStartLabelOffset = (): number => {
    const startMarkerOffset = this.calculateMarkerStartOffset();
    const startLabelHalf = this.state.labelStartWidth / 2;

    return startMarkerOffset < startLabelHalf
      ? 0
      : startMarkerOffset - startLabelHalf;
  };

  isBelowMaxPadding = (value: number, gap: number = 0): boolean => {
    const maxPadding = this.getMaxPadding(gap);
    return value + gap < maxPadding;
  };

  setPaddingForTwoLabels = (): void => {
    const startLabelOffset = this.getStartLabelOffset();
    const endMarkerOffset = this.calculateMarkerEndOffset();
    const endLabelHalf = this.state.labelEndWidth / 2;
    const endLabelOffset =
      endMarkerOffset < endLabelHalf ? 0 : endMarkerOffset - endLabelHalf;

    const isBelowMaxPadding = this.isBelowMaxPadding(
      startLabelOffset + endLabelOffset,
      10,
    );

    const hasOffsetChanged =
      this.state.paddingLeft !== startLabelOffset ||
      this.state.paddingRight !== endLabelOffset;

    if (isBelowMaxPadding && hasOffsetChanged) {
      this.setState({
        paddingLeft: startLabelOffset,
        paddingRight: endLabelOffset,
      });
    }

    requestAnimationFrame(this.setPaddingForTwoLabels);
  };

  setPaddingForOneLabel = (): void => {
    const startLabelOffset = this.getStartLabelOffset();
    const isBelowMaxPadding = this.isBelowMaxPadding(startLabelOffset);
    const hasOffsetChanged = this.state.paddingLeft !== startLabelOffset;

    if (isBelowMaxPadding) {
      if (hasOffsetChanged) {
        this.setState({
          paddingLeft: startLabelOffset,
        });
      }

      if (this.state.labelStartAtMax) {
        this.setState({ labelStartAtMax: false });
      }
    } else if (this.state.labelStartAtMax === false) {
      this.setState({ labelStartAtMax: true });
    }

    requestAnimationFrame(this.setPaddingForOneLabel);
  };

  saveFullWidth = (e: OnLayout) => {
    this.setState({ width: Math.floor(e.nativeEvent.layout.width) });
  };

  saveLabelStartWidth = (e: OnLayout) => {
    this.setState({ labelStartWidth: Math.floor(e.nativeEvent.layout.width) });
  };

  saveLabelEndWidth = (e: OnLayout) => {
    this.setState({ labelEndWidth: Math.floor(e.nativeEvent.layout.width) });
  };

  render() {
    return (
      <View
        style={[
          styles.sliderLabels,
          {
            paddingLeft: this.state.paddingLeft,
            paddingRight: this.state.paddingRight,
            justifyContent: this.state.labelStartAtMax
              ? 'flex-end'
              : 'space-between',
          },
        ]}
        onLayout={this.saveFullWidth}
      >
        <View onLayout={this.saveLabelStartWidth}>
          <Text style={styles.label}>{this.props.startLabel}</Text>
        </View>
        {this.props.endLabel && (
          <View onLayout={this.saveLabelEndWidth}>
            <Text style={styles.label}>{this.props.endLabel}</Text>
          </View>
        )}
      </View>
    );
  }
}

const styles = StyleSheet.create({
  sliderLabels: {
    width: '100%',
    flexDirection: 'row',
    marginTop: 10,
    marginBottom: 5,
  },
  label: {
    fontSize: 14,
    color: defaultTokens.paletteBlueNormal,
  },
});
