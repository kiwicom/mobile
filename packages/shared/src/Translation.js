// @flow

import * as React from 'react';
import {
  CaseTransform,
  getTranslation,
  type SupportedTransformationsType,
  type TranslationKeysType,
} from '@kiwicom/mobile-localization';

import Text from './Text';

type CommonProps = {|
  testID?: string,
  textTransform?: SupportedTransformationsType,
|};

type Props =
  | {|
      id: TranslationKeysType,
      values?: Object,
      ...CommonProps,
    |}
  | {|
      passThrough: ?string | ?number, // may be undefined because of optional chaining
      ...CommonProps,
    |};

/**
 * Regular translation using registered keys:
 *
 * ```js
 * <Translation id="SingleHotel.Rating.Stars" />
 * ```
 *
 * So called "pass through" translation (this props is not being translated
 * but passed through the component).
 *
 * ```js
 * <Translation passThrough="★★★★★" />
 * ```
 *
 * It's possible to apply text transformations in both cases:
 *
 * ```js
 * <Translation passThrough="★★★★★" textTransform="uppercase" />
 * <Translation id="SingleHotel.Rating.Stars" textTransform="uppercase" />
 * ```
 */

// use this property to highlight translations for screenshoting or debugging
const highlightTranslations = false;
export default function Translation(props: Props) {
  const containerStyle = highlightTranslations
    ? {
        backgroundColor: props.id ? 'orange' : 'pink',
      }
    : undefined;
  if (props.id === undefined && props.passThrough != null) {
    return (
      <Text style={containerStyle}>
        {/* $FlowExpectedError: we do not allow to use 'string' in the 'Text' components but translations are exceptions. */}
        {CaseTransform(String(props.passThrough), props.textTransform)}
      </Text>
    );
  } else if (props.id === undefined) {
    return null;
  }

  const values = props.values || {};
  return (
    <Text style={containerStyle}>
      {/* $FlowExpectedError: we do not allow to use 'string' in the 'Text' components but translations are exceptions. */}
      {CaseTransform(getTranslation(props.id, values), props.textTransform)}
    </Text>
  );
}
