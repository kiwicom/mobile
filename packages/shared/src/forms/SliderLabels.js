// @flow

import * as React from 'react';
import { defaultTokens } from '@kiwicom/mobile-orbit';
import { View } from 'react-native';

import type { OnLayout } from '../../types/Events';
import Text from '../Text';
import StyleSheet from '../PlatformStyleSheet';
import Price from '../Price';
import type { TranslationType } from '../../types/Translation';
import useCalculateSliderPositions from './useCalculateSliderPositions';

type Props = {|
  +startLabel: TranslationType | React.Element<typeof Price>,
  +startValue: number,
  +endLabel?: TranslationType | React.Element<typeof Price>,
  +endValue?: number,
  +max: number,
  +min: number,
|};

export default function SliderLabels(props: Props) {
  const [width, setWidth] = React.useState(0);
  const [labelStartWidth, setLabelStartWidth] = React.useState(0);
  const [labelStartAtMax, setLabelStartAtMax] = React.useState(false);
  const [labelEndWidth, setLabelEndWidth] = React.useState(0);
  const [paddingLeft, setPaddingLeft] = React.useState(0);
  const [paddingRight, setPaddingRight] = React.useState(0);

  useCalculateSliderPositions({
    labelEndWidth,
    labelStartAtMax,
    labelStartWidth,
    paddingLeft,
    paddingRight,
    props,
    width,
    setPaddingLeft,
    setLabelStartAtMax,
    setPaddingRight,
    min: props.min,
    max: props.max,
    endValue: props.endValue,
    startValue: props.startValue,
  });

  function saveFullWidth(e: OnLayout) {
    setWidth(Math.floor(e.nativeEvent.layout.width));
  }

  function saveLabelStartWidth(e: OnLayout) {
    setLabelStartWidth(Math.floor(e.nativeEvent.layout.width));
  }

  function saveLabelEndWidth(e: OnLayout) {
    setLabelEndWidth(Math.floor(e.nativeEvent.layout.width));
  }

  return (
    <View
      style={[styles.sliderLabels, labelStartAtMax && styles.sliderLabelsEnd]}
      onLayout={saveFullWidth}
      testID="sliderLabelsContainer"
    >
      <View
        onLayout={saveLabelStartWidth}
        style={{
          transform: [{ translateX: labelStartAtMax ? 0 : paddingLeft }],
        }}
        testID="startLabelContainer"
      >
        <Text style={styles.label}>{props.startLabel}</Text>
      </View>
      {props.endLabel && (
        <View
          onLayout={saveLabelEndWidth}
          style={{
            transform: [{ translateX: -paddingRight }],
          }}
          testID="endLabelContainer"
        >
          <Text style={styles.label}>{props.endLabel}</Text>
        </View>
      )}
    </View>
  );
}

const styles = StyleSheet.create({
  sliderLabels: {
    width: '100%',
    flexDirection: 'row',
    marginTop: 10,
    marginBottom: 5,
    justifyContent: 'space-between',
  },
  sliderLabelsEnd: {
    justifyContent: 'flex-end',
  },
  label: {
    fontSize: 14,
    color: defaultTokens.paletteBlueNormal,
  },
});
