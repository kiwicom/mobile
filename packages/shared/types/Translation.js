// @flow

import * as React from 'react';

import Translation from '../src/Translation';
import TranslationFragment from '../src/TranslationFragment';

export type TranslationType =
  | React.Element<typeof Translation>
  | React.Element<typeof TranslationFragment>;
