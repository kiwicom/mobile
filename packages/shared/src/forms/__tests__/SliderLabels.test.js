// @flow strict

import * as React from 'react';
import renderer from 'react-test-renderer';
import ShallowRenderer from 'react-test-renderer/shallow';

import Translation from '../../Translation';
import SliderLabels from '../SliderLabels';

const shallowRenderer = new ShallowRenderer();

it('renders with one label', () => {
  expect(
    shallowRenderer.render(
      <SliderLabels
        startLabel={<Translation passThrough="label" />}
        startValue={46}
        max={1000}
        min={1}
      />,
    ),
  ).toMatchSnapshot();
});

it('renders with two labels', () => {
  expect(
    shallowRenderer.render(
      <SliderLabels
        startLabel={<Translation passThrough="start label" />}
        startValue={46}
        endLabel={<Translation passThrough="end label" />}
        endValue={850}
        max={1000}
        min={1}
      />,
    ),
  ).toMatchSnapshot();
});

it('should have startValue in the middle', () => {
  const wrapper = renderer.create(
    <SliderLabels
      max={5}
      min={1}
      startLabel={<Translation passThrough="label" />}
      startValue={3}
    />,
  );
  const viewContainer = wrapper.root.findByProps({
    testID: 'sliderLabelsContainer',
  });
  const startLabel = wrapper.root.findByProps({
    testID: 'startLabelContainer',
  });

  renderer.act(() => {
    viewContainer.props.onLayout({
      nativeEvent: { layout: { width: 500 } },
    });
    startLabel.props.onLayout({ nativeEvent: { layout: { width: 20 } } });
  });

  expect(startLabel.props.style).toMatchInlineSnapshot(`
        Object {
          "transform": Array [
            Object {
              "translateX": 290,
            },
          ],
        }
    `);
});

it('should display the label at the end', () => {
  const wrapper = renderer.create(
    <SliderLabels
      max={5}
      min={1}
      startLabel={<Translation passThrough="label" />}
      startValue={5}
    />,
  );

  const viewContainer = wrapper.root.findByProps({
    testID: 'sliderLabelsContainer',
  });
  const startLabel = wrapper.root.findByProps({
    testID: 'startLabelContainer',
  });

  renderer.act(() => {
    viewContainer.props.onLayout({
      nativeEvent: { layout: { width: 500 } },
    });
    startLabel.props.onLayout({ nativeEvent: { layout: { width: 20 } } });
  });

  expect(startLabel.props.style).toMatchInlineSnapshot(`
    Object {
      "transform": Array [
        Object {
          "translateX": 472,
        },
      ],
    }
  `);
});

it('should have translateX on both labels', () => {
  const wrapper = renderer.create(
    <SliderLabels
      max={20}
      min={1}
      startLabel={<Translation passThrough="label" />}
      startValue={8}
      endLabel={<Translation passThrough="label" />}
      endValue={16}
    />,
  );

  const viewContainer = wrapper.root.findByProps({
    testID: 'sliderLabelsContainer',
  });
  const startLabel = wrapper.root.findByProps({
    testID: 'startLabelContainer',
  });
  const endLabel = wrapper.root.findByProps({
    testID: 'endLabelContainer',
  });

  renderer.act(() => {
    viewContainer.props.onLayout({ nativeEvent: { layout: { width: 500 } } });
    startLabel.props.onLayout({ nativeEvent: { layout: { width: 20 } } });
    endLabel.props.onLayout({ nativeEvent: { layout: { width: 20 } } });
  });

  expect(startLabel.props.style).toMatchInlineSnapshot(`
        Object {
          "transform": Array [
            Object {
              "translateX": 190,
            },
          ],
        }
    `);
  expect(endLabel.props.style).toMatchInlineSnapshot(`
        Object {
          "transform": Array [
            Object {
              "translateX": -90,
            },
          ],
        }
    `);
});

it('should have no translateX', () => {
  const wrapper = renderer.create(
    <SliderLabels
      max={100}
      min={1}
      startLabel={<Translation passThrough="label" />}
      startValue={1}
      endLabel={<Translation passThrough="label" />}
      endValue={100}
    />,
  );

  const viewContainer = wrapper.root.findByProps({
    testID: 'sliderLabelsContainer',
  });
  const startLabel = wrapper.root.findByProps({
    testID: 'startLabelContainer',
  });
  const endLabel = wrapper.root.findByProps({
    testID: 'endLabelContainer',
  });

  renderer.act(() => {
    viewContainer.props.onLayout({ nativeEvent: { layout: { width: 500 } } });
    startLabel.props.onLayout({ nativeEvent: { layout: { width: 20 } } });
    endLabel.props.onLayout({ nativeEvent: { layout: { width: 20 } } });
  });

  expect(startLabel.props.style.transform[0].translateX).toBe(0);
  expect(endLabel.props.style.transform[0].translateX).toBe(0);
});
