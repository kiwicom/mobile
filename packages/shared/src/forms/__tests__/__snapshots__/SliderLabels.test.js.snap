// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders with one label 1`] = `
<View
  onLayout={[Function]}
  style={
    Array [
      Object {
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginBottom": 5,
        "marginTop": 10,
        "width": "100%",
      },
      false,
    ]
  }
  testID="sliderLabelsContainer"
>
  <View
    onLayout={[Function]}
    style={
      Object {
        "transform": Array [
          Object {
            "translateX": 0,
          },
        ],
      }
    }
    testID="startLabelContainer"
  >
    <Text
      style={
        Object {
          "color": "#0176D2",
          "fontSize": 14,
        }
      }
    >
      <Translation
        passThrough="label"
      />
    </Text>
  </View>
</View>
`;

exports[`renders with two labels 1`] = `
<View
  onLayout={[Function]}
  style={
    Array [
      Object {
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginBottom": 5,
        "marginTop": 10,
        "width": "100%",
      },
      false,
    ]
  }
  testID="sliderLabelsContainer"
>
  <View
    onLayout={[Function]}
    style={
      Object {
        "transform": Array [
          Object {
            "translateX": 0,
          },
        ],
      }
    }
    testID="startLabelContainer"
  >
    <Text
      style={
        Object {
          "color": "#0176D2",
          "fontSize": 14,
        }
      }
    >
      <Translation
        passThrough="start label"
      />
    </Text>
  </View>
  <View
    onLayout={[Function]}
    style={
      Object {
        "transform": Array [
          Object {
            "translateX": -0,
          },
        ],
      }
    }
    testID="endLabelContainer"
  >
    <Text
      style={
        Object {
          "color": "#0176D2",
          "fontSize": 14,
        }
      }
    >
      <Translation
        passThrough="end label"
      />
    </Text>
  </View>
</View>
`;
